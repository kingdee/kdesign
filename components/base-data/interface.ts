import { tuple } from '../_utils/type'

export const Modes = tuple('multiple', 'single')
export type Mode = typeof Modes[number]

export type SelectValue = string | string[] | number | number[] | IOptionsProps[] | null

export interface IAdvancedSelectorProps<T = SelectValue> {
  prefixCls?: string // 样式前缀
  className?: string // 样式名
  children?: React.ReactNode
  style?: React.CSSProperties // 内联样式
  disabled?: boolean // 禁用
  dropdownStyle?: React.CSSProperties // 下拉样式
  notFoundContent?: React.ReactNode // 列表为空时文本
  placeholder?: string // 选择框默认文字
  showFrequent?: boolean // 是否显示历史记录与我的收藏面板
  showCollectIcon?: boolean // 是否显示收藏按钮
  showDetailIcon?: boolean // 是否显示详情按钮
  suffixIcon?: React.ReactNode // 选择框后缀图标
  mode?: Mode // 模式
  value?: IOptionsProps[] | IOptionsProps // 指定选中的条目
  columns?: [] // 下拉列表的展示列
  options?: [] // 下拉列表的数据
  size?: number // 下拉列表显示数量
  optionLabelProp?: string // 回填到选择框的属性
  dropdownFooterRender?: React.ReactNode // 自定义下拉框footer内容
  visible?: boolean
  collectList?: [] // 收藏列表
  historyList?: [] // 我的历史
  loading?: boolean // 加载中状态
  searchField?: string // 搜索字段 默认 all（全部字段）
  onChange?: () => void // 选中option调用此函数
  onSearch?: (val: string) => void // input的value变化时调用
  onSelect?: (value: T extends (infer I)[] ? I : T, option: React.ReactElement<any>) => void
  onShowDetail?: (value: IOptionsProps[]) => void
  onShowMore?: () => void
  onCollect?: (flag: boolean, value: IOptionsProps) => void
}

export interface IIndxPosListProps {
  end: number
  start: number
}

export interface IColumnsProps {
  title?: string
  key: string
}

export interface IOptionsProps {
  value?: string | number
  label?: string
}

export interface ISearchInfoProps {
  searchValue?: string
  previousInputValue: string
  previousEditValue?: string
  searchIndex?: number
  selectionStart?: number
  deleteEndIndx: number
  editOptions?: any[]
}
