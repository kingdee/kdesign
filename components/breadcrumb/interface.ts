import { ReactNode } from 'react'
import { DropDownProps } from '../dropdown'
import { tuple } from '../_utils/type'

export const ColorModelTypes = tuple('emphasize', 'weaken')
export type ColorModelType = typeof ColorModelTypes[number]

export interface IBreadcrumbProps {
  style?: Record<string, unknown> // 内联样式
  className?: string // 类名
  colorModel?: ColorModelType // 颜色模式，默认是 'emphasize'
  prefixCls?: string // 样式前缀
  items: Array<IBreadcrumbItem> // 面包屑导航的各项，支持传入 ReactNode 类型
  separator?: ReactNode // 分隔符，默认是 '/'
  children?: ReactNode // 子元素
  onItemClick?: (item: IBreadcrumbItem, index: number) => void // 点击面包屑导航的回调函数
}

export interface IBreadcrumbItems {
  item: IBreadcrumbItem
  index: number
  separator?: ReactNode
  openEllipsis?: boolean
  onItemClick?: (item: IBreadcrumbItem, index: number) => void // 点击面包屑导航的回调函数
}

export interface IBreadcrumbItem {
  className?: string // 类名
  title?: ReactNode
  dropdownProps?: DropDownProps
  href?: string // 链接地址，不能和path共用
  path?: string // 拼接地址，不能和href共用
  icon?: ReactNode
}

export interface IItemsWidth {
  width: number
  index: number
}
