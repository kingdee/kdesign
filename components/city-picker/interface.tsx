import * as React from 'react'
import { tuple } from '../_utils/type'

export type City = {
  id?: string | number
  country?: string
  province?: string
  name?: string
  [key: string]: any
}

export const BorderTypes = tuple('none', 'underline', 'bordered')
export type BorderType = typeof BorderTypes[number]

export const Sizes = tuple('large', 'middle', 'small') // 选择框大小
export type Size = typeof Sizes[number]

export type Type = 'domestic' | 'foreign'

export type CityList = Array<City>

export interface CityPickerProps {
  prefixCls?: string // 样式前缀
  size?: Size // 尺寸
  borderType?: BorderType // 边框类型
  notFoundContent?: React.ReactNode // 列表为空时文本
  showArrow?: boolean // 显示下拉小箭头
  defaultOpen?: boolean // 默认是否展开下拉列表
  allowClear?: boolean // 允许清除内容
  placeholder?: React.ReactNode // 选择框默认文字
  type?: Type
  loading?: boolean
  domesticList?: CityList
  foreignList?: CityList
  commonList?: CityList
  className?: string
  disabled?: boolean
  style?: React.CSSProperties
  title?: string
  clearIcon?: React.ReactNode // 清除图标
  suffixIcon?: React.ReactNode // 选择框后缀图标
  optionHighlightProps?: string // 搜索时 根据该属性值高亮城市名
  value?: string | number
  defaultValue?: string | number
  description?: React.ReactNode
  showDescription?: boolean
  onChange?: (value: string | number, city: City) => void
  onBlur?: (e: React.ChangeEvent<HTMLSpanElement>) => void
  onFocus?: () => void
  onSearch?: (value: string) => void
  onClear?: () => void
  itemRender?: (city: City) => React.ReactNode
  onTabPaneChange?: (id: Type) => void
}

export interface ICityPickerOptionProps {
  disabled?: boolean
  city?: City
  value?: string | number
  children?: React.ReactNode
  className?: string
  style?: React.CSSProperties
  onChangeSelect?: (city: City | null) => void
  renderCityInfo?: (city: City, flag?: boolean, symbol?: string) => React.ReactNode
  itemRender?: (city: City) => React.ReactNode
}
