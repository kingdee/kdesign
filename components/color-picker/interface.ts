import { tuple } from '../_utils/type'
import { PopperProps } from '../_utils/usePopper'

export const ColorTypes = tuple('HEX', 'HSB', 'RGB', 'HSL', 'themeColor')
export type ColorType = typeof ColorTypes[number]
export interface IColorTypesObj {
  type: ColorType
  value: string
}

export const BorderTypes = tuple('underline', 'bordered')
export type BorderType = typeof BorderTypes[number]
export interface IColorPickerProps extends PopperProps {
  // 设置颜色选择器的值，可以是十六进制颜色值，RGB颜色值，HSL颜色值，或者颜色名称
  value: string
  // 设置颜色选择器的默认值，可以是十六进制颜色值，RGB颜色值，HSL颜色值，或者颜色名称
  defaultValue: string
  // 选择器输入框类名
  className: string
  // 选择输入框样式
  style: React.CSSProperties
  // 输入框类型
  borderType: BorderType
  // 输入框右侧的图标
  suffixIcon: (rgbColor: string) => React.ReactNode
  // 功能色名称
  functionalColorName: string
  // 功能色色值
  functionalColor: string
  // 功能色开关的名称
  switchName: { name: string; internationalName?: string }
  // 是否跟随功能色,定义了功能色才能控制显示跟随主题色开关
  showSwitch: boolean
  // 是否展示色彩格式转换框
  showColorTransfer: boolean
  // 是否展示系统预设颜色
  showPresetColor: boolean
  // 占位字符
  placeholder: string
  // 用户自定义的预设颜色
  presetColor: string[]
  // 是否展示拾色容器
  showColorPickerBox: { showBox: boolean; showHue: boolean; showOpacity: boolean }
  // 颜色选择面板显隐配置
  defaultOpen: boolean
  // 当颜色选择器的值发生变化时触发的回调函数
  onChange: (inputValue: string, formatColorArr: Array<IColorTypesObj>) => void
}

export interface IColorPickerPanelProps {
  alpha: number
  alphaNoVerifyVal: string
  clickedColorIndex?: number
  colTypeArr: Array<IColorTypesObj>
  correctColorValue: string
  currentColorType: IColorTypesObj['type']
  functionalColor?: string
  functionalColorName: string
  switchName: { name: string; internationalName?: string }
  inputRef: any
  isFollow: boolean
  presetColor?: string[]
  showSwitch?: boolean
  showColorTransfer?: boolean
  showPresetColor?: boolean
  showColorPickerBox?: { showBox?: boolean; showHue?: boolean; showOpacity?: boolean }
  showColorPickerPanel: boolean
  value: string
  setCurrentColorType: (currentColorType: IColorTypesObj['type']) => void
  setColTypeArr: (colTypeArr: Array<IColorTypesObj>) => void
  setClickedColorIndex: (clickedColorIndex: number) => void
  setIsFollow: (isFollow: boolean) => void
  setAlphaNoVerifyVal: (NoVerifyVal: string) => void
  setAlpha: (alpha: number) => void
  setCorrectColorValue: (colorValue: string) => void
  setInputColorValue: (colorValue: string) => void
  setShowPanel: (showPanel: boolean) => void
  onChange?: (inputValue: string, formatColorArr: Array<IColorTypesObj>) => void
}
