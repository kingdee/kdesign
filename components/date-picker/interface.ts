import { BorderType, InputSiteType } from '../input/input'
import { DisabledDataProps } from './date-picker'

export type WeekTitleType = [string, string, string, string, string, string, string]
export type MonthTitleType = [
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
  string,
]

export interface Locale {
  // locale: string
  'DatePicker.placeholder': string
  'DatePicker.yearPlaceholder': string
  'DatePicker.quarterPlaceholder': string
  'DatePicker.monthPlaceholder': string
  'DatePicker.weekPlaceholder': string
  'DatePicker.timePlaceholder': string
  'DatePicker.now': string
  'DatePicker.confrim': string
  'DatePicker.today': string
  'DatePicker.year': string
  'DatePicker.month': string
  'DatePicker.weekTitle': WeekTitleType
  'DatePicker.monthTitle': MonthTitleType
  'DatePicker.rangePlaceholder': [string, string]
  'DatePicker.rangeYearPlaceholder': [string, string]
  'DatePicker.rangeMonthPlaceholder': [string, string]
  'DatePicker.rangeWeekPlaceholder': [string, string]
  'DatePicker.rangeQuarterPlaceholder': [string, string]
  'DatePicker.rangeTimePlaceholder': [string, string]
}

export interface InnerLocale {
  locale: string
  placeholder: string
  yearPlaceholder: string
  quarterPlaceholder: string
  monthPlaceholder: string
  weekPlaceholder: string
  timePlaceholder: string
  now: string
  confrim: string
  today: string
  year: string
  month: string
  weekTitle: WeekTitleType
  monthTitle: MonthTitleType
  rangePlaceholder: [string, string]
  rangeYearPlaceholder: [string, string]
  rangeMonthPlaceholder: [string, string]
  rangeWeekPlaceholder: [string, string]
  rangeQuarterPlaceholder: [string, string]
  rangeTimePlaceholder: [string, string]
}

export type InnerLocaleKey = keyof InnerLocale

export type PanelMode = 'time' | 'date' | 'week' | 'month' | 'quarter' | 'year' | 'decade'

export type PickerMode = Exclude<PanelMode, 'datetime' | 'decade'>

export type DateType = Date

export type CustomFormat = (value: DateType) => string

export type DateFormat = string

export type DateFormats = Array<string>

export type NullableDateType = DateType | null | undefined

export type OnSelect = (value: DateType, type: 'key' | 'mouse' | 'submit') => void
export interface DisabledTimes {
  disabledHours?: () => number[]
  disabledMinutes?: (hour: number) => number[]
  disabledSeconds?: (hour: number, minute: number) => number[]
}

export type DisabledTime = (date: DateType | null) => DisabledTimes

export type OnPanelChange = (value: DateType, mode: PanelMode) => void

export type EventValue = DateType | null
export type RangeValue = [EventValue, EventValue] | null

export type RangeList = {
  label: string
  onClick: () => void
  onMouseEnter: () => void
  onMouseLeave: () => void
}[]

export interface Components {
  button?: React.ComponentType | string
  rangeItem?: React.ComponentType | string
}

export interface PanelSharedProps {
  value?: NullableDateType
  defaultPickerValue?: DateType
  locale: InnerLocale
  disabledDate?: DisabledDataProps

  prevIcon?: React.ReactNode
  nextIcon?: React.ReactNode
  superPrevIcon?: React.ReactNode
  superNextIcon?: React.ReactNode

  // 面板操作（焦点操作、键盘操作）next version
  // operationRef: React.MutableRefObject<PanelRefProps>

  // onPanelChange: (mode: PanelMode | null, viewValue: DateType) => void
}

export type DateRender = (currentDate: DateType, today: DateType) => React.ReactNode

export type CellRender = (currentDate: DateType) => React.ReactNode

export type RangeType = 'start' | 'end'

export interface RangeInfo {
  range: RangeType
}

export type RangeDateRender = (currentDate: DateType, today: DateType, info: RangeInfo) => React.ReactNode

export interface SharedTimeProps extends DisabledTimes {
  format?: string
  showNow?: boolean
  showHour?: boolean
  showMinute?: boolean
  showSecond?: boolean
  use12Hours?: boolean
  hourStep?: number
  minuteStep?: number
  secondStep?: number
  hideDisabledOptions?: boolean
  defaultValue?: DateType
}

export interface OutInputCommonProps {
  format: DateFormat
  picker: PickerMode
  id?: string
  autoFocus?: boolean
  allowClear?: boolean
  className?: string
  borderType?: BorderType
  style?: React.CSSProperties
  size?: InputSiteType
  readOnly?: boolean
  suffixIcon?: React.ReactNode
  clearIcon?: React.ReactNode

  onFocus?: React.FocusEventHandler<HTMLInputElement>
  onBlur?: React.FocusEventHandler<HTMLInputElement>
  onMouseDown?: React.MouseEventHandler<HTMLDivElement>
  onMouseUp?: React.MouseEventHandler<HTMLDivElement>
  onMouseEnter?: React.MouseEventHandler<HTMLDivElement>
  onMouseLeave?: React.MouseEventHandler<HTMLDivElement>
  onClick?: React.MouseEventHandler<HTMLDivElement>
  onContextMenu?: React.MouseEventHandler<HTMLDivElement>
}

export interface InputCommonProps {
  panelDivRef: React.RefObject<HTMLDivElement>
  needConfirmButton: boolean
  open: boolean
  prefixCls: string
  locale: InnerLocale
  dataOrAriaProps?: Record<string, any>
  status?: 'error'
}

export interface OutPopperProps {
  dropdownClassName?: string
  popupStyle?: React.CSSProperties
  showArrow?: boolean
}

export interface TimeUnit {
  label: React.ReactText
  value: number
  disabled: boolean
}
