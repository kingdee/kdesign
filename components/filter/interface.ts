import React from 'react'
import { Menu } from '../dropdown/menu'
import { IQuickSearchProps } from '../search'
import { DirectionType } from '../config-provider/ConfigContext'

type TabKey = 'condition' | 'scheme'

export type TFilterValue = {
  [key: string]: string[] | IScheme
}

export type TOption = {
  value?: string
  label?: string
  component?: React.ReactNode
}

export interface ICondition {
  key: string
  label: string
  options: TOption[]
  required?: boolean
  multiple?: boolean
  addition?: React.ReactNode
}

export interface IField {
  name: string
  label: string
  required?: boolean
  operators: React.ReactElement
  assignOptions: Record<string, React.ReactElement>
}

export interface ISchemeItem {
  fieldName: string
  value: string
  operator: string
  required?: boolean
}

export interface IScheme {
  name?: string
  preset?: boolean
  items: ISchemeItem[]
}

export interface IFilterProps {
  spread?: boolean
  fields?: IField[]
  className?: string
  schemes?: IScheme[]
  value?: TFilterValue
  defaultTabKey?: TabKey
  defaultSpread?: boolean
  title?: React.ReactNode
  conditions?: ICondition[]
  search?: IQuickSearchProps
  style?: React.CSSProperties
  defaultValue?: TFilterValue
  onSchemeSave?: (scheme: IScheme) => void
  onSpreadChange?: (spread: boolean) => void
  getSchemeSettingMenu?: (scheme: IScheme) => typeof Menu
  onChange?: (value: TFilterValue, condition?: ICondition, option?: TOption) => void
}

export interface IConditionProps {
  tab: string
  prefixCls: string
  value: TFilterValue
  conditions: ICondition[]
  FilterLangMsg: Record<string, any>
  onConditionRemove: (key: string, e: React.MouseEvent) => void
  onConditionChange: (value: TFilterValue, condition?: ICondition, option?: TOption) => void
}

export interface ISchemeProps {
  tab: string
  fields: IField[]
  prefixCls: string
  schemes: IScheme[]
  direction: DirectionType
  value: { scheme: IScheme }
  FilterLangMsg: Record<string, any>
  onSchemeSave: (scheme: IScheme) => void
  onSchemeSearch: (scheme: IScheme) => void
  getSchemeSettingMenu?: (scheme: IScheme) => void
}
