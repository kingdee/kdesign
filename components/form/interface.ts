// field name path
export type NamePath = string | number

// form layout, default is `vertical`
export type FormLayout = 'horizontal' | 'vertical' | 'inline'

// label text align, default is `left`
export type LabelAlign = 'left' | 'right'

// Store
export type StoreValue = any
export interface Store {
  [name: string]: StoreValue
}

export interface FieldError {
  [name: string]: string
}

export interface Fields {
  [name: string]: { current: FieldInstance }
}

// FormInstance, use for external
export interface FormInstance<Values = any> {
  getFieldError: (name: NamePath) => string | undefined
  getFieldsError: (namePathList?: NamePath[]) => FieldError
  getFieldValue: (name: NamePath) => any
  getDefaultValue: (name: NamePath) => any
  getFieldsValue: (namePathList?: NamePath[]) => any
  resetFields: (fields?: NamePath[]) => void
  setFieldsValue: (value: Values) => void
  setFieldValue: (name: NamePath, value: Values) => void
  validateFields: (namePathList?: NamePath[]) => Promise<Values>
  submit: () => void
  getInternalHooks: (secret: string) => InternalHooks | null
}

export interface InnerLocale {
  locale: string
  requiredMessage: string
}

// Internal form instance
export interface InternalFormInstance extends FormInstance {
  labelWidth?: string | number
  labelAlign?: LabelAlign
  wrapperWidth?: string | number
  vertical?: boolean
  disabled?: boolean
  local?: InnerLocale
  getInternalHooks: (key: string) => InternalHooks | null
}

// Field instance
export type Stores = {
  prev: Store
  curr: Store
}
export interface FieldInstance {
  onStoreChange: (stores: Stores, namePathList: NamePath[] | null, source: NotifySource) => void
  meta: {
    name?: NamePath
    rules?: Rule[]
    defaultValue?: any
    trigger?: string | string[]
  }
}

// Form Instance callbacks
export interface Callbacks<Values = any> {
  onValuesChange?: (changedValues: any, values: Values) => void
  onFinish?: (value: Values) => void
  onFinishFailed?: (errorInfo: ValidateErrorEntity<Values>) => void
}

// Actions
export interface UpdateAction {
  type: 'updateValue'
  namePath: NamePath
  value: StoreValue
}

export interface ValidateAction {
  type: 'validateField'
  namePath: NamePath
}

export type ReducerAction = UpdateAction | ValidateAction

// Internal hooks, use for every field
export interface InternalHooks {
  dispatch: (action: ReducerAction) => void
  setDefaultValues: (values: Store) => void
  setCallbacks: (callbacks: Callbacks) => void
  registerField: (name: NamePath, field: { current: FieldInstance }) => void
  deleteField: (name: NamePath) => void
}

export type NotifySource =
  | 'externalUpdateValue' // 外部调用方法更新字段值
  | 'updateValue' // 内部更新字段值
  | 'reset' // 重置字段值
  | 'validateField' // 校验
  | 'validateFinish' // 校验结束

// export type FieldNotifyInfo = NotifyInfo & {
//   store: Store
//   errorMessages: FieldError
// }

export interface ValidateErrorEntity<Values = any> {
  values: Values
  errorFields: { name: NamePath; errors: string[] }[]
}

// Rule
type Validator = (rule: RuleObject, value: StoreValue, callback: (error?: string) => void) => Promise<void | any> | void
export type RuleType =
  | 'string'
  | 'number'
  | 'boolean'
  | 'method'
  | 'regexp'
  | 'integer'
  | 'float'
  | 'object'
  | 'enum'
  | 'date'
  | 'url'
  | 'hex'
  | 'email'

export interface ValidatorRule {
  message?: string | React.ReactElement
  validator: Validator
}
interface BaseRule {
  enum?: StoreValue[]
  len?: number
  max?: number
  message?: string | React.ReactElement
  min?: number
  pattern?: RegExp
  required?: boolean
  transform?: (value: StoreValue) => StoreValue
  type?: RuleType
  whitespace?: boolean

  /** Customize rule level `validateTrigger`. Must be subset of Field `validateTrigger` */
  validateTrigger?: string | string[]
}
type AggregationRule = BaseRule & Partial<ValidatorRule>
interface ArrayRule extends Omit<AggregationRule, 'type'> {
  type: 'array'
  defaultField?: RuleObject
}
export type RuleRender = (form: FormInstance) => RuleObject
export type RuleObject = AggregationRule | ArrayRule
export type Rule = RuleObject | RuleRender
