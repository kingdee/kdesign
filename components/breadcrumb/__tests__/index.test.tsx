import { mount, render } from 'enzyme'
import React from 'react'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'
import { ColorModelTypes, IBreadcrumbItem } from '../interface'
import Breadcrumb from '../index'
import Icon from '../../icon'

const config = [
  {
    title: 'KDesign',
    href: 'https://kingdee.design/',
  },
  {
    title: 'KDesign React',
    href: 'https://react.kingdee.design/',
  },
  {
    title: 'Breadcrumb',
  },
]

const iconConfig = [
  {
    title: 'KDesign',
    href: 'https://kingdee.design/',
    icon: <Icon type="star" />,
  },
  {
    title: 'KDesign React',
    href: 'https://react.kingdee.design/',
  },
  {
    title: 'Breadcrumb',
  },
]

const errorConfig = [
  {
    title: 'KDesign',
    href: 'https://kingdee.design/',
    path: 'kd/main',
  },
  {
    title: 'KDesign React',
    href: 'https://react.kingdee.design/',
    path: 'kd/main',
  },
  {
    title: 'Breadcrumb',
  },
]

describe('Breadcrumb', () => {
  // 1. mount test
  describe('1. mount test', () => {
    mountTest(Breadcrumb)
    ColorModelTypes.forEach((type) => {
      mountTest(() => <Breadcrumb items={config} colorModel={type} />)
    })
  })

  // 2. render test
  describe('2. render test', () => {
    it('renders correctly', () => {
      expect(render(<Breadcrumb items={config} />)).toMatchSnapshot()

      ColorModelTypes.forEach((type) => {
        expect(render(<Breadcrumb items={config} colorModel={type} />)).toMatchSnapshot()
      })
    })
  })

  // 3. warns in component
  describe('3. warns in component', () => {
    it('warns if href and path coexist', () => {
      const mockWarn = jest.fn()
      jest.spyOn(console, 'warn').mockImplementation(mockWarn)
      const wrapper = mount(<Breadcrumb items={errorConfig} />)
      wrapper.find('.kd-breadcrumb-item').at(0).simulate('click')
      expect(mockWarn).toHaveBeenCalledTimes(1)
      expect(mockWarn.mock.calls[0][0]).toMatch(
        'Warning: [kdesign]-breadcrumb: `href` and `path` not coexist  within the `item` object',
      )
    })
  })

  // 4. render null or undefined without errors
  describe('4. render null or undefined without errors', () => {
    it('render null or undefined without errors', () => {
      const wrapper = (
        <Breadcrumb items={config}>
          {null}
          {undefined}
        </Breadcrumb>
      )
      expect(wrapper).toMatchSnapshot()
    })
  })

  // 5. displayName
  describe('5. displayName', () => {
    it('should have displayName static property', () => {
      expect(Breadcrumb.displayName).toBe('Breadcrumb')
    })
  })

  // 6. class state
  describe('6. class state', () => {
    it('other properties should be passed', () => {
      const wrapper = mount(
        <Breadcrumb items={config} className="my-class" data-test="test">
          hello world
        </Breadcrumb>,
      )
      expect(wrapper).toHaveClassName('my-class')
      expect(wrapper.prop('data-test')).toEqual('test')

      const defaultBreadcrumb = mount(<Breadcrumb items={config} />)
      expect(defaultBreadcrumb.find('.kd-breadcrumb-item')).toHaveLength(3)
      expect(defaultBreadcrumb.find('.kd-breadcrumb-item').at(0).childAt(0)).toHaveClassName('kd-breadcrumb-item-text')
      expect(defaultBreadcrumb.find('.kd-breadcrumb-item').at(0).childAt(1)).toHaveClassName(
        'kd-breadcrumb-item-separator',
      )
      expect(defaultBreadcrumb.find('.kd-breadcrumb-item-link')).toExist()
      expect(defaultBreadcrumb.find('.kd-breadcrumb-item').at(0)).toHaveClassName('kd-breadcrumb-item-emphasize-model')
      expect(defaultBreadcrumb.find('.kd-breadcrumb-item').at(2)).toHaveClassName(
        'kd-breadcrumb-item-emphasize-model-current',
      )

      const weakenBreadcrumb = mount(<Breadcrumb items={config} colorModel="weaken" />)
      expect(weakenBreadcrumb.find('.kd-breadcrumb-item')).toHaveLength(3)
      expect(weakenBreadcrumb.find('.kd-breadcrumb').childAt(0)).not.toHaveClassName('kd-breadcrumb-item-link')
      expect(weakenBreadcrumb.find('.kd-breadcrumb-item').at(0)).toHaveClassName('kd-breadcrumb-item-weaken-model')
      expect(weakenBreadcrumb.find('.kd-breadcrumb-item').at(2)).toHaveClassName(
        'kd-breadcrumb-item-weaken-model-current',
      )

      const iconBreadcrumb = mount(<Breadcrumb items={iconConfig} />)
      expect(iconBreadcrumb.find('.kd-breadcrumb-item').at(0).childAt(0)).toHaveClassName('kd-breadcrumb-item-icon')
    })
  })

  // 7.component interaction(event)
  describe('7.component interaction(event)', () => {
    it('should trigger onItemClick callback', () => {
      const onClick = jest.fn()
      const wrapper = mount(<Breadcrumb items={config} onItemClick={onClick} />)
      wrapper.find('.kd-breadcrumb').childAt(0).simulate('click')
      expect(onClick).toHaveBeenCalledTimes(1)
    })
  })

  // 8.config provider
  describe('8.config provider', () => {
    it('should config use config provider', () => {
      const clipboardConfig = {
        compDefaultProps: {
          Breadcrumb: {
            colorModel: 'weaken',
            separator: '>',
          },
        },
      }
      const wrapper = mount(
        <ConfigProvider value={clipboardConfig}>
          <Breadcrumb items={config} />
        </ConfigProvider>,
      )
      expect(wrapper.find('.kd-breadcrumb-item').at(0)).toHaveClassName('kd-breadcrumb-item-weaken-model')
      expect(wrapper.find('.kd-breadcrumb-item-separator').at(0)).toHaveText('>')
    })
  })

  // 9. ref test

  // 10. api test
  describe('10. api test', () => {
    it('separator api', () => {
      const wrapper = mount(<Breadcrumb separator="X" items={config} />)
      expect(wrapper.find('.kd-breadcrumb-item-separator').at(0)).toHaveText('X')
    })
    it('items api', () => {
      const itemsConfig = [
        {
          title: 'KDesign',
          className: 'my-item',
          href: 'https://kingdee.design/',
          icon: '*',
          dropdownProps: {
            defaultVisible: true,
            menu: [
              {
                key: 1,
                label: '按钮',
                href: 'https://react.kingdee.design/components/button',
              },
              {
                key: 2,
                label: '剪贴板',
                href: 'https://react.kingdee.design/components/clipboard',
              },
              {
                key: 3,
                label: '图标',
                href: 'https://react.kingdee.design/components/icon',
              },
            ],
          },
        },
      ]
      const wrapper = mount(<Breadcrumb items={itemsConfig as unknown as IBreadcrumbItem[]} />)
      expect(wrapper.find('.kd-breadcrumb-item-icon')).toHaveText('*')
      expect(wrapper.find('.kd-breadcrumb-item-text')).toHaveText('KDesign')
      expect(wrapper.find('.kd-breadcrumb-item')).toHaveClassName('my-item')
      expect(wrapper.find('.kd-breadcrumb-item').find('.kd-dropdown-trigger')).toExist()
    })
  })

  // 11. special case
  describe('11. special case', () => {
    it('should hide section items to adaptive parent width', () => {
      const adaptiveConfig = [
        {
          title: 'KDesign',
          href: 'https://kingdee.design/',
        },
        {
          title: 'KDesign React',
          href: 'https://react.kingdee.design/',
        },
        {
          title: 'KDesign React',
          href: 'https://react.kingdee.design/',
        },
        {
          title: 'Breadcrumb',
        },
      ]
      const wrapper = mount(
        <div className="parent">
          <Breadcrumb items={adaptiveConfig} />
        </div>,
      )
      wrapper.setProps({ style: { width: '200px' } })
      wrapper.update()
      expect(wrapper.find('.kdicon-more').length).toEqual(1)
    })
  })
})
