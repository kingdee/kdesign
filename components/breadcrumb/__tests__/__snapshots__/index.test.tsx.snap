// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb 2. render test renders correctly 1`] = `
Array [
  <div
    class="kd-breadcrumb"
  >
    <div
      class="kd-breadcrumb-item kd-breadcrumb-item-link kd-breadcrumb-item-emphasize-model"
    >
      <div
        class="kd-breadcrumb-item-text"
        style="max-width:unset;overflow:inherit"
        title="KDesign"
      >
        KDesign
      </div>
      <span
        class="kd-breadcrumb-item-separator"
      >
        /
      </span>
    </div>
    <div
      class="kd-breadcrumb-item kd-breadcrumb-item-link kd-breadcrumb-item-emphasize-model"
    >
      <div
        class="kd-breadcrumb-item-text"
        style="max-width:unset;overflow:inherit"
        title="KDesign React"
      >
        KDesign React
      </div>
      <span
        class="kd-breadcrumb-item-separator"
      >
        /
      </span>
    </div>
    <div
      class="kd-breadcrumb-item kd-breadcrumb-item-emphasize-model-current"
    >
      <div
        class="kd-breadcrumb-item-text"
        style="max-width:unset;overflow:inherit"
        title="Breadcrumb"
      >
        Breadcrumb
      </div>
    </div>
  </div>,
  <div
    class="kd-breadcrumb-hide-icon"
  >
    <i
      class="kdicon kdicon-more"
    />
    <span
      class="kd-breadcrumb-item-separator"
    >
      /
    </span>
  </div>,
]
`;

exports[`Breadcrumb 2. render test renders correctly 2`] = `
Array [
  <div
    class="kd-breadcrumb"
  >
    <div
      class="kd-breadcrumb-item kd-breadcrumb-item-link kd-breadcrumb-item-emphasize-model"
    >
      <div
        class="kd-breadcrumb-item-text"
        style="max-width:unset;overflow:inherit"
        title="KDesign"
      >
        KDesign
      </div>
      <span
        class="kd-breadcrumb-item-separator"
      >
        /
      </span>
    </div>
    <div
      class="kd-breadcrumb-item kd-breadcrumb-item-link kd-breadcrumb-item-emphasize-model"
    >
      <div
        class="kd-breadcrumb-item-text"
        style="max-width:unset;overflow:inherit"
        title="KDesign React"
      >
        KDesign React
      </div>
      <span
        class="kd-breadcrumb-item-separator"
      >
        /
      </span>
    </div>
    <div
      class="kd-breadcrumb-item kd-breadcrumb-item-emphasize-model-current"
    >
      <div
        class="kd-breadcrumb-item-text"
        style="max-width:unset;overflow:inherit"
        title="Breadcrumb"
      >
        Breadcrumb
      </div>
    </div>
  </div>,
  <div
    class="kd-breadcrumb-hide-icon"
  >
    <i
      class="kdicon kdicon-more"
    />
    <span
      class="kd-breadcrumb-item-separator"
    >
      /
    </span>
  </div>,
]
`;

exports[`Breadcrumb 2. render test renders correctly 3`] = `
Array [
  <div
    class="kd-breadcrumb"
  >
    <div
      class="kd-breadcrumb-item kd-breadcrumb-item-link kd-breadcrumb-item-weaken-model"
    >
      <div
        class="kd-breadcrumb-item-text"
        style="max-width:unset;overflow:inherit"
        title="KDesign"
      >
        KDesign
      </div>
      <span
        class="kd-breadcrumb-item-separator"
      >
        /
      </span>
    </div>
    <div
      class="kd-breadcrumb-item kd-breadcrumb-item-link kd-breadcrumb-item-weaken-model"
    >
      <div
        class="kd-breadcrumb-item-text"
        style="max-width:unset;overflow:inherit"
        title="KDesign React"
      >
        KDesign React
      </div>
      <span
        class="kd-breadcrumb-item-separator"
      >
        /
      </span>
    </div>
    <div
      class="kd-breadcrumb-item kd-breadcrumb-item-weaken-model-current"
    >
      <div
        class="kd-breadcrumb-item-text"
        style="max-width:unset;overflow:inherit"
        title="Breadcrumb"
      >
        Breadcrumb
      </div>
    </div>
  </div>,
  <div
    class="kd-breadcrumb-hide-icon"
  >
    <i
      class="kdicon kdicon-more"
    />
    <span
      class="kd-breadcrumb-item-separator"
    >
      /
    </span>
  </div>,
]
`;

exports[`Breadcrumb 4. render null or undefined without errors render null or undefined without errors 1`] = `
<Breadcrumb
  items={
    Array [
      Object {
        "href": "https://kingdee.design/",
        "title": "KDesign",
      },
      Object {
        "href": "https://react.kingdee.design/",
        "title": "KDesign React",
      },
      Object {
        "title": "Breadcrumb",
      },
    ]
  }
/>
`;
